libvirt/uri/ssh.go

```diff
@@ -91,74 +125,184 @@ func (u *ConnectionURI) dialSSH() (net.Conn, error) {
+			skip_verify = false
+		}
 	}
 
 	if knownHostsPath == "" {
-		knownHostsPath = defaultSSHKnownHostsPath
+		knownHosts, err := sshcfg.Get(target, "UserKnownHostsFile")
+		if err == nil && knownHosts != "" {
+			knownHostsPath = knownHosts
+		} else {
+			knownHostsPath = defaultSSHKnownHostsPath
+		}
 	}
 
 	hostKeyCallback := ssh.InsecureIgnoreHostKey()
-	if doVerify {
-		cb, err := knownhosts.New(os.ExpandEnv(knownHostsPath))
+	hostKeyAlgorithms := []string{ // https://github.com/golang/go/issues/29286
+		// this can be solved using https://github.com/skeema/knownhosts/tree/main
+		// there is an open issue requiring attention
+		ssh.KeyAlgoED25519,
+		ssh.KeyAlgoRSA,
+		ssh.KeyAlgoRSASHA256,
+		ssh.KeyAlgoRSASHA512,
+		ssh.KeyAlgoSKECDSA256,
+		ssh.KeyAlgoSKED25519,
+		ssh.KeyAlgoECDSA256,
+		ssh.KeyAlgoECDSA384,
+		ssh.KeyAlgoECDSA521,
+	}
+	if !skip_verify {
+		kh, err := knownhosts.New(os.ExpandEnv(knownHostsPath))
 		if err != nil {
 			return nil, fmt.Errorf("failed to read ssh known hosts: %w", err)
 		}
-		hostKeyCallback = cb
-	}
+		log.Printf("[DEBUG] Using known hosts file '%s' for target '%s'", os.ExpandEnv(knownHostsPath), target)
 
-	username := u.User.Username()
-	if username == "" {
-		sshu, err := sshcfg.Get(u.Host, "User")
-		log.Printf("[DEBUG] SSH User: %v", sshu)
-		if err != nil {
-			log.Printf("[DEBUG] ssh user: system username")
-			u, err := user.Current()
+		hostKeyCallback = func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+			err := kh(net.JoinHostPort(hostName, port), remote, key)
 			if err != nil {
-				return nil, fmt.Errorf("unable to get username: %w", err)
+				log.Printf("Host key verification failed for host '%s' (%s) %v: %v", hostName, remote, key, err)
 			}
-			sshu = u.Username
+			return err
 		}
-		username = sshu
+
+		keyAlgs, err := sshcfg.Get(target, "HostKeyAlgorithms")
+		if err == nil && keyAlgs != "" {
+			log.Printf("Got host key algorithms '%s'", keyAlgs)
+			hostKeyAlgorithms = strings.Split(keyAlgs, ",")
+		}
+
 	}
 
 	cfg := ssh.ClientConfig{
-		User:            username,
+		User:            u.User.Username(),
 		HostKeyCallback: hostKeyCallback,
-		Auth:            authMethods,
+		HostKeyAlgorithms: hostKeyAlgorithms,
 		Timeout:         dialTimeout,
 	}
 
-	port := u.Port()
-	if port == "" {
-		port = defaultSSHPort
+	proxy, err := sshcfg.Get(target, "ProxyCommand")
+	if err == nil && proxy != "" {
+		log.Printf("[WARNING] unsupported ssh ProxyCommand '%v'", proxy)
 	}
 
-	sshClient, err := ssh.Dial("tcp", fmt.Sprintf("%s:%s", u.Hostname(), port), &cfg)
-	if err != nil {
-		return nil, err
+	proxy, err = sshcfg.Get(target, "ProxyJump")
+	var bastion *ssh.Client
+	if err == nil && proxy != "" {
+		log.Printf("[DEBUG] found ProxyJump '%v'", proxy)
+
+		// this is a proxy jump: we recurse into that proxy
+		bastion, err = u.dialHost(proxy, sshcfg, depth + 1)
+		if err != nil {
+			return nil, fmt.Errorf("failed to connect to bastion host '%v': %w", proxy, err)
+		}
 	}
```

Comment:
If the err != nil, it is silently ignored instead of returning. My suggestion: if err != nil, return, then only condition on proxy being non empty to execute the recursion.

Reply:
Unclear, can you rephrase please? Did you mean "if the error == nil" ?

Are you perhaps saying there can be a condition in which both `bastion` and `err` are nil?

Reply:
What I mean is that `proxy, err = sshcfg.Get(target, "ProxyJump")` can return error, and in that case you would continue with the flow, instead of returning the function. Yes you would skip dialing the proxy (assuming it returned empty string in addition to the error), but the flow would continue anyway.

Wouldn't it make more sense to return early if `proxy, err = sshcfg.Get(target, "ProxyJump")` returns an error. And then you execute the code that dials the proxy only testing for `proxy != ""` ?

Reply:
`ProxyJump` can be absent, in which case `Get()` (as opposed to `GetStrict()`)returns empty string [as per sshconfig docs](https://github.com/kevinburke/ssh_config/blob/1d09c0b50564c4a7f8c56c9d5d6d935e06ee94da/config.go#L113-L118). This isn't an error per-se as it simply means there's no proxy. This covers case 1 of the conditional on line 250, which is `err == nil` and `proxy == ""`. In this case, the flow should proceed without proxy.

The second case is if the sshconfig file had a parse error (i.e. `err != nil`) and we don't care about `proxy`. In this case, it means that the sshconfig file couldn't be parsed. Expected behaviour in those circumstances (e.g. when using ssh from cli) is to simply honor the command itself and ignore any ssh_config directives altogether, which in this case corresponds to a flow-through without Proxy.

> Wouldn't it make more sense to return early if proxy, err = sshcfg.Get(target, "ProxyJump") returns an error. And then you execute the code that dials the proxy only testing for proxy != "" ?

I think there's a case to be made that (maybe) the very first successful call to `sshcfg.Get()` guarantees that every subsequent call is successful. However, I'd not feel comfortable breaking the abstraction and depending on this behaviour: it would make the code very convoluted and introduce race conditions (i.e. if `ssh_config` gets touched during this setup).

Thoughts?

Reply:
@dmacvicar see above for my thoughts on this codepath.